WIP: Libraryfy Lomap

richardjgowers opened the pull request:
This PR tries to make Lomap usable as a library (rather than solely as a command line tool).

Changes made:
- tests changed to use pytest (to allow parametrisation)
- tests updated to use `pkg_resources` to make them more portable
- `DBMolecules` no longer calls `argparse`, instead just takes arguments like a more normal class
- added `lomap.dbmol._startup_inner` as a non-command line way of invoking lomap (`lomap.dbmol.startup` (the CLI entrypoint) now just parses command line args and proceeds to this inner function)

There's quite a few tests failing, but I think these were failing before I started tinkering, (maybe due to recent changes?) so some archaeology is required to try and restore the original regression tests / correct the desired behaviour.

Also it's not clear if the PR should go into this repo, or if OFE will be "adopting"/forking Lomap going forwards.

richardjgowers committed:
use pkg_resources for test file locations

richardjgowers committed:
add dbmol._startup_inner to offer library-like access

update tests to use this entrypoint, rather than hacky sys.argv hacking

richardjgowers committed:
dunderscores are annoying and unnecessary

richardjgowers committed:
fix DBMolecules.__add__

richardjgowers committed:
remove use of argparse inside class __init__...

DBMolecules.options now a dict not a namespace

richardjgowers committed:
switch to use pytest

for parametrization

mikemhenry committed:
first pass on CI

mikemhenry committed:
Merge pull request #1 from mikemhenry/feat/setup_GHA_CI

first pass on CI

mikemhenry committed:
drats forgot the env file

davidlmobley commented:
I suspect everything was broken here, so you probably haven't broken anything in bringing this in. I also don't have anyone on my end ready to review, so I'd be happy to have you take over this repo entirely if you want to own it (and you can appoint your own reviewer). Grab me via my calendly if you'd like to talk directly about it.

The other thing that might/might not be relevant is that the Cresset/Edinburgh/BioSimSpace folks are using a fork of this already (and we recently brought their stuff back in) so it's somewhat up to you as to whether you find a way to share infrastructure with them or take this in a different direction.

jmichel80 commented:
Hiya, yes it would be good to keep @JenkeScheen in the loop so we can all still keep working with the same LOMAP version

mikemhenry committed:
Merge branch 'OpenFreeEnergy:main' into main

mikemhenry committed:
pin boost and simplify matrix

mikemhenry committed:
change the way we install the package

mikemhenry committed:
doh just use -e

mikemhenry committed:
update readme with build badge

mikemhenry committed:
don't think we need to point pytest to the test file

davidlmobley commented:
@jmichel80 do you think he'd want to review PRs?

JenkeScheen commented:
Hi folks, @davidlmobley yes I'd be happy to. LOMAP has been part of my daily research for a several months now so would be nice to be involved.

richardjgowers commented on lomap/dbmol.py:
```diff
@@ -120,90 +120,65 @@ def __init__(self, directory, parallel=1, verbose='off',
             # logging.basicConfig(format='%(levelname)s:\t%(message)s', level=logging.DEBUG)
 
-        if __name__ == '__main__':
-            self.options = parser.parse_args()
```
@JenkeScheen this is the only breaking change.  If you're using this I can probably put it back in, but I thought it was a little ugly.  The other change is that `.options` is a dict not a namespace, so access is a little different, but items have the same name.  If that breaks a lot of things for you I could hack together a quasi dict that allows `.attribute` access

JenkeScheen replied:
I'm currently using the cresset fork of LOMAP that lives in [BioSimSpace/Align](https://github.com/michellab/BioSimSpace/tree/devel/python/BioSimSpace/Align/_lomap) so my current work isn't affected by changes made in the Mobley repo (most of my work has been in a function that pulls data from LOMAP, not changing it, see [BSS.Align.generateNetwork()](https://github.com/michellab/BioSimSpace/blob/devel/python/BioSimSpace/Align/_align.py#L82))

JenkeScheen commented on lomap/dbmol.py:
```diff
@@ -331,13 +306,13 @@ def read_molecule_files(self):
         # The .mol2 and .sdf file formats are the only supported so far
-        mol_fnames = glob.glob(self.options.directory + "/*.mol2")
-        mol_fnames += glob.glob(self.options.directory + "/*.sdf")
+        mol_fnames = glob.glob(self.options['directory'] + "/*.mol2")
```
There's an issue here in the way that LOMAP handles molecules: generating networks with the same series of ligands in both .sdf and .mol2 (switching through e.g. obabel) formats results in different graph outputs (i.e. different sets of suggested edges). Anecdotally I've observed that .sdf-based graphs are better (i.e. lead to ultimately more accurate FE predictions) than .mol2-based graphs, likely due to RDKit handling the molecule information in a different way. 

Out of scope for this PR I think but would be good to deal with at some point soon.

richardjgowers replied:
I think in general having an API based on files (not molecular representations like rdkit) isn't a great idea, but it would be too much of a change to make.

JenkeScheen replied:
Agreed - I don't know about OFE but for direct users of LOMAP at the very least a log warning that discourages the use of .mol2 inputs would be a reasonable idea for the time being.

JenkeScheen commented on lomap/graphgen.py:
```diff
@@ -996,10 +996,10 @@ def write_graph(self, output_no_images, output_no_graph):
         log = 'The following files have been generated:'
         if not output_no_graph:
-            log += f'\n{self.dbase.options.name}.dot\tGraph file'
+            log += f'\n{self.dbase.options["name"]}.dot\tGraph file'
```
should it not be ['name'] over ["name"] to conform to style?

richardjgowers replied:
it's already inside a f-string, so you'd have to swap both, but not really a big deal either way

JenkeScheen replied:
ah, missed that, check

JenkeScheen commented on test/test_lomap.py:
```diff
@@ -122,27 +121,24 @@ def test_get_set_add(self):
+        assert (s_strict == p_strict).all()
+        assert (s_loose == p_loose).all()
 
     def test_read_mol2_files(self):
```
this could be a good spot for an additional test to check between .sdf and .mol2 handling.

JenkeScheen commented on test/test_lomap.py:
```diff
@@ -122,27 +121,24 @@ def test_get_set_add(self):
-
     # Test which heterocycles I can grow (growing off a phenyl)
     # Test by Max indicates that growing complex heterocycles tends
     # to fail, so only allow growing phenyl, furan and pyrrole
```
These comments can be removed, it's a Cresset artefact and likely very SOMD (or single topology-)specific. There should be some possibility for users to add customisation like this for themselves.

JenkeScheen commented on test/test_lomap.py:
```diff
@@ -118,68 +152,68 @@ def test_read_mol2_files(self):
+            self.assertEqual(MC.heterocycles_rule(penalty=4), expected,'Fail on heterocycle check for ' + fn)
 
     # Tests by Max indicate that growing a sulfonamide all in one go is 
     # dodgy, so disallow it
```
also internal Cresset notes, feel free to remove

JenkeScheen commented on test/test_lomap.py:
```diff
@@ -209,138 +243,151 @@ def test_clip_on_3d(self):
-            comp=Chem.MolFromMolFile('test/transforms/'+d[1],sanitize=False, removeHs=False)
+        for fn1, fn2, expected in testdata:
+            parent=Chem.MolFromMolFile(fn1, sanitize=False, removeHs=False)
+            comp=Chem.MolFromMolFile(fn2, sanitize=False, removeHs=False)
```
I may have missed the definition of `fn`/`fn1/2`, but perhaps it would be better to replace with `f_name` or something similarly more verbose.

JenkeScheen reviewed (changes requested):
Added some comments on small changes, everything else looks good to me. I'm not experienced with implementing CI so for this part an additional reviewer could be more helpful, but looking at it everything seems to be in order.

There are quite a few Cresset rules present in the code; I think it would be OK to leave these in at this stage but would be worth a discussion on where/how to move them (as different users will be using different FEP engines requiring different rules).

mikemhenry committed:
lets see if this fixes 'test_complete_run'

richardjgowers committed:
use pytest tmpdirs for tests

richardjgowers committed:
removed unused & broken function

richardjgowers committed:
changed MCS to use kwargs not argparse namespace

internally also stored as dict rather than argparse namespace object

richardjgowers committed:
removed dunderscores

richardjgowers committed:
fixup

richardjgowers committed:
Code was relicensed to MIT a while ago, updated some references

richardjgowers committed:
delete duplicate test file

these tests are all in test_lomap

richardjgowers committed:
fix ring size rule tests

these were changed a while ago to return 0.1 but the tests were left...

also some pep8 inside this rule

JenkeScheen commented on lomap/mcs.py:
```diff
@@ -526,37 +512,37 @@ def set_ring_counter(mol):
-                                    ringMatchesRingOnly=True,
-                                    completeRingsOnly=True,
-                                    matchChiralTag=False)
+        __mcs = rdFMCS.FindMCS([self._moli_noh, self._molj_noh],
```
`__mcs` still has dunderscore here, is that intentional?

richardjgowers replied:
The class dunderscores do that annoying name mangling in Python, this isn't a class attribute so it's less egregious. Just walking the line between keeping the diff minimal and making the code more usable.

JenkeScheen commented on devtools/conda-recipe/meta.yaml:
```diff
@@ -48,6 +48,6 @@ test:
-  license: LGPL
+  license: MIT
   license_file: LICENCE.txt
   summary: Alchemical free-energy calculations planner
```
could be worth referencing the original paper url as well, e.g. https://link.springer.com/article/10.1007%2Fs10822-013-9678-y

JenkeScheen commented on lomap/graphgen.py:
```diff
@@ -26,18 +26,6 @@
 # and Christopher Summa at Summa Lab, Dept. of Computer Science,
 # University of New Orleans and it has just been adapded to the new Lomap code
```
adapded -> adapted

richardjgowers committed:
remove isclose from tests

pytest.approx gives better traceback, (isclose squashes the result to a bool)

richardjgowers committed:
some pep8

removed some print statements

yoinked some functions from inside methods

richardjgowers committed:
fixed up failing test

split out different rules tests

added skip to tests for now...

mikemhenry committed:
Merge branch 'main' into main

lohedges commented:
Thanks for your work on this. LOMAP is currently the only vendored package within BioSImSpace so it would be fantastic if there was a working conda-installable version that we could use instead. Let me know if you need any input from our end.

Cheers.

JenkeScheen reviewed (commented):
checked these commits, all changes are sensible.

mikemhenry committed:
just test rdkit 2021 and python 3.8 and 3.9

mikemhenry committed:
lets see if removing the boost pin works

richardjgowers committed:
Merge pull request #2 from mikemhenry/main

richardjgowers committed:
started MCS regression tests

richardjgowers committed:
toluene basic regression tests

richardjgowers committed:
use fixture of fixtures to avoid repetition

richardjgowers committed:
Merge pull request #3 from OpenFreeEnergy/MCS_regression_tests

Mcs regression tests

richardjgowers committed:
lift atom_hybridization function into mcs

IAlibay committed:
fix mapping

IAlibay committed:
Cleanup

IAlibay committed:
Fix function access

IAlibay committed:
fix pep8

richardjgowers committed:
added mcs tests for explicit H mappings

richardjgowers committed:
Merge pull request #7 from OpenFreeEnergy/test-remap

fix explicit H mapping

richardjgowers committed:
update README to reflect status

richardjgowers committed:
updated requirements in README

deleted stale travis file

richardjgowers committed:
flipped CHANGELOG upside down

richardjgowers committed:
updated CHANGELOG for v2.0.0

JenkeScheen commented:
Is a review for this PR still needed, considering https://github.com/OpenFreeEnergy/Lomap/issues/6? Seems I might be slightly out of the loop.

On another note, I'd be happy to do a quick check that generated networks are identical between lomap1 and lomap2 at some point; I have code in place that aggregates edges for (almost) all congeneric series in https://github.com/openforcefield/protein-ligand-benchmark/tree/master/data using lomap1, it shouldn't be too difficult to compare with lomap2.. Especially with the new conda-forge release! (we can continue the discussion on this elsewhere if needed).

richardjgowers commented:
@JenkeScheen yes sorry, we ended up keeping our fork of lomap (aka `lomap2` on conda-forge) and we'll be maintaining it from there.  You should be able to run your checks against the conda-forge package, that's live now.

davidlmobley commented:
Do you guys want to do a PR here to update the README.md to point to that as the maintained version, or should I do it?
